### PR TITLE
CVSL-1438 Refactoring structure of conditions

### PR DESCRIPTION
**Pull request by andrewrlee**
Simplifying condition structure by flattening out instances.
This also allows us also extract out a MEZ condition type - so we can remove file upload from conditions that don't have uploads.

Note the swagger ui doesn't cope with discriminators very well - hoping code gen tools handle it better. 

The generate example doesn't show the discriminator for some reason:
<kbd>
Screenshot 2023-10-27 at 11 50 48
</kbd>

 but when you expand the model it shows it:
<kbd>
Screenshot 2023-10-27 at 11 51 00
</kbd>

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/ElectronicMonitoringAdditionalCondition.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.resource.publicApi.model.licence.additionalConditions
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/ElectronicMonitoringAdditionalCondition.kt:1]: <kbd>Screenshot 2023-10-27 at 11 46 27</kbd>

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/MultipleExclusionZoneAdditionalCondition.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.resource.publicApi.model.licence.additionalConditions
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/MultipleExclusionZoneAdditionalCondition.kt:1]: <kbd>
Screenshot 2023-10-27 at 11 47 39
</kbd>

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.resource.publicApi.model.licence.additionalConditions
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt:1]: <kbd>
Screenshot 2023-10-27 at 11 48 34
</kbd>

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      schema = ElectronicMonitoringAdditionalCondition::class,
+    ),
+    DiscriminatorMapping(
+      value = ConditionTypes.MULTIPLE_EXCLUSION_ZONE,
+      schema = MultipleExclusionZoneAdditionalCondition::class,
+    ),
+  ],
+)
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+sealed interface AdditionalCondition {
```
Comment (gurps-singh) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt:29]: So we are saying that each of the discriminator versions of additional conditions (i.e. `ElectronicMonitoringAdditionalCondition`) implements this interface but certain additional conditions will have their own extra properties?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      value = ConditionTypes.ELECTRONIC_MONITORING,
+      schema = ElectronicMonitoringAdditionalCondition::class,
+    ),
+    DiscriminatorMapping(
+      value = ConditionTypes.MULTIPLE_EXCLUSION_ZONE,
+      schema = MultipleExclusionZoneAdditionalCondition::class,
+    ),
+  ],
+)
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
```
Comment (gurps-singh) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt:28]: How does `@JsonTypeInfo` being declared on the interface here relate to the consequent declarations for each of the additional conditions type (e.g. `@JsonTypeName(ConditionTypes.ELECTRONIC_MONITORING)` for the `ElectronicMonitoringAdditionalCondition`)

**Review by gurps-singh — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt (hunk 4 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/publicApi/model/licence/additionalConditions/AdditionalCondition.kt:29]: yep thats right

**Pull request merged**